wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationService.swift

```diff
@@ -135,10 +136,56 @@ public final class ConversationService: ConversationServiceInterface {
         )
     }
 
+    private func internalCreateGroupWithRetryIfNeeded(
+        teamID: UUID?,
+        name: String?,
+        users: Set<ZMUser>,
+        accessMode: ConversationAccessMode,
+        accessRoles: Set<ConversationAccessRoleV2>,
+        enableReceipts: Bool,
+        messageProtocol: WireDataModel.MessageProtocol,
+        completion: @escaping (Swift.Result<ZMConversation, ConversationCreationFailure>) -> Void) {
+
+            func retry(excludingDomains domains: Set<String>) {
+                let usersToExclude = users.belongingTo(domains: domains)
+                let reachableUsers = Set(users).subtracting(usersToExclude)
+
+                if !reachableUsers.isEmpty {
+                    internalCreateGroupConversation(
+                        teamID: teamID,
+                        name: name,
+                        users: reachableUsers,
+                        failedToAddUsers: usersToExclude,
+                        accessMode: accessMode,
+                        accessRoles: accessRoles,
+                        enableReceipts: enableReceipts,
+                        messageProtocol: messageProtocol,
+                        completion: completion)
+                }
+            }
+
+            internalCreateGroupConversation(
+                teamID: teamID,
+                name: name,
+                users: users,
+                accessMode: accessMode,
+                accessRoles: accessRoles,
+                enableReceipts: enableReceipts,
+                messageProtocol: messageProtocol) { result in
+                    switch result {
+                    case .failure(.networkError(.unreachableDomains(let domains))):
+                        retry(excludingDomains: domains)
+                    default:
+                        completion(result)
+                    }
+                }
+        }
+
```

Comment:
This could be simplified as follows. This approach also avoids having to pass in the `failedToAddUsers` as an argument, which I think is better because it we keep the logic of dealing with unreachable users in the local scope (as opposed to moving it to another method):

```swift
    private func internalCreateGroupWithRetryIfNeeded(
        teamID: UUID?,
        name: String?,
        users: Set<ZMUser>,
        accessMode: ConversationAccessMode,
        accessRoles: Set<ConversationAccessRoleV2>,
        enableReceipts: Bool,
        messageProtocol: WireDataModel.MessageProtocol,
        completion: @escaping (Swift.Result<ZMConversation, ConversationCreationFailure>) -> Void
    ) {
        func createGroup(
            withUsers users: Set<ZMUser>,
            completion: @escaping (Swift.Result<ZMConversation, ConversationCreationFailure>) -> Void
        ) {
            internalCreateGroupConversation(
                teamID: teamID,
                name: name,
                users: users,
                accessMode: accessMode,
                accessRoles: accessRoles,
                enableReceipts: enableReceipts,
                messageProtocol: messageProtocol,
                completion: completion
            )
        }

        createGroup(withUsers: users) { result in
            switch result {
            case .failure(.networkError(.unreachableDomains(let domains))):
                let unreachableUsers = users.belongingTo(domains: domains)
                let reachableUsers = Set(users).subtracting(unreachableUsers)

                createGroup(withUsers: reachableUsers) { retryResult in
                    if case .success(let conversation) = retryResult {
                        conversation.appendFailedToAddUsersSystemMessage(
                            users: unreachableUsers,
                            sender: .selfUser(in: self.context),
                            at: Date()
                        )
                    }

                    completion(retryResult)
                }

            default:
                completion(result)
            }
        }
    }
```

What do you think?